### PR TITLE
filesystem based isRecordStale

### PR DESCRIPTION
**Pull request by brianPlummer**
- filesystem based isRecordStale
- staleAwarePersister that is expiration aware
- flags for storeBuilder to specify a cache policy UNSPECIFIED, REFRESH_ON_STALE, NETWORK_BEFORE_STALE

### CODE REVIEW

File: filesystem/src/main/java/com/nytimes/android/external/fs/StaleAwarePersister.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    @Inject
+    public StaleAwarePersister(FileSystem fileSystem,
+                               @Nonnull TimeUnit expirationUnit,
+                               long expirationDuration) {
+        super(fileSystem);
+        this.expirationDuration = expirationDuration;
+        this.expirationUnit = expirationUnit;
+    }
+
+    public RecordState isRecordStale(@Nonnull BarCode barCode) {
```
Comment (brianPlummer) [inline filesystem/src/main/java/com/nytimes/android/external/fs/StaleAwarePersister.java]: should i call it StalePersister?

File: filesystem/src/main/java/com/nytimes/android/external/fs/filesystem/FileSystem.java [hunk 2]
```diff
@@ -92,4 +95,17 @@
      * @return exists, duh
      */
     boolean exists(String file);
+
+    /**
+     * compares age of file with given expiration time and returns
+     * appropriate recordState
+     *
+     * @param expirationUnit
```
Comment (brianPlummer) [inline filesystem/src/main/java/com/nytimes/android/external/fs/filesystem/FileSystem.java]: will do

File: filesystem/src/test/java/com/nytimes/android/external/fs/StaleAwarePersisterTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    @Test
+    public void readExists() throws FileNotFoundException {
+        when(fileSystem.exists(SourcePersister.pathForBarcode(simple)))
+                .thenReturn(true);
+        when(fileSystem.read(SourcePersister.pathForBarcode(simple))).thenReturn(bufferedSource);
+
+        BufferedSource returnedValue = sourcePersister.read(simple).toBlocking().single();
+        assertThat(returnedValue).isEqualTo(bufferedSource);
```
Comment (brianPlummer) [inline filesystem/src/test/java/com/nytimes/android/external/fs/StaleAwarePersisterTest.java]: ??? restate please. this test was primarily copied from the sourcePersister.  i didn't know if we wanted to duplicate the entire test suite to ensure that everything worked as expected, or just test the new methods i implemented.

File: store/src/main/java/com/nytimes/android/external/store/base/impl/StoreBuilder.java [hunk 4]
```diff
@@ -70,6 +77,17 @@
         return this;
     }
 
+    public StoreBuilder<Raw> refreshOnStale() {
```
Comment (brianPlummer) [inline store/src/main/java/com/nytimes/android/external/store/base/impl/StoreBuilder.java]: will do

File: filesystem/src/test/java/com/nytimes/android/external/fs/StaleAwarePersisterTest.java (hunk 3 above)
Comment (digitalbuddha) [inline filesystem/src/test/java/com/nytimes/android/external/fs/StaleAwarePersisterTest.java]: only new methods.  You're making a mock for exists within this test yet nothing is calling it. I think you can remove the entire test

**Comment by digitalbuddha** [comment]
will reopen with a bit more. The changes could not be checked in alone

**Comment by PaulWoitaschek** [comment]
Why is this barcode based? Shouldn't there be an abstraction to resolve a generic type to a path?

**Comment by digitalbuddha** [comment]
That's the impl for our middleware. There will be interfaces  so that you can create your own persister impl with stale awareness ☺️ writing tests will get this reopened once done.

**Pull request closed without merging**